### PR TITLE
Add assets to binary

### PR DESCRIPTION
**Pull request by Wind0r**
go-bindata add all assets to the binary with the help from go generate

if the folder assets is available the server will use it else files from binary will be used

### CODE REVIEW

File: server/assets.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+//go:generate go-bindata -pkg main -ignore .../.DS_Store -o files.go assets/...
```
Comment (mrnugget) [inline server/assets.go]: Is there another way to ignore files than to hardcode the name into this file?

**Comment by mrnugget** [comment]
Great idea!

I'm wondering: this does add the hard dependency of `github.com/jteeuwen/go-bindata`, right? Could you  explain that in the README?

And shouldn't `files.go` be in `.gitignore`? It's autogenerated on every build, right?

**Comment by Wind0r** [comment]
go generated need to be executed by hand so files.go need to be included, else the repository will not be go get/install able.
Normaly i would not include them but in go we need to, but github is greate and ignores  files with "// Code generated by" (go generate,protobufs etc)

It should not have any impact. If applikatoni found the folder "assets" it will use the folder else it will use the files from the binary.

I did this as a first step to a single binary, but doing it for the migration scripts need some more time.
Also i want to show what i meant with fancy solution.

**Comment by mrnugget** [comment]
Thanks for the explanation.

I'm still not sure if I want to merge this. Not that you didn't do a good job. What makes me hesitate:

1. New dependency and therefore a new hurdle when trying to set up a local dev environment
2. Local development of assets is now more complicated: without this it was super easy to modify the assets and reload the page in the browser. Now there's a need to regenerate the binary.
3. A new build step with `go generate`

All of these make playing around with and developing Applikatoni a lot more complicated with (at the moment) little payoff.

**Comment by Wind0r** [comment]
If there is a way to bundle goose with the binary i would like to get this merged.

1. The dependency is "only" needed if someone change the assets  
2. Local files still are preferred, so it is still super easy. [See assets.go ](https://github.com/Wind0r/applikatoni/blob/6c603e18db9012a3d52d08301a5850f4cd005d60/server/assets.go#L15)
3. I only added it to be save that the new assets are included. It's not needed.

**Comment by mrnugget** [comment]
Thanks for the explanation. I think I now better understand how it works.

But in regards to 1.: How does this work? In order to use `go generate` the dependency is needed right? And without `go generate` the `assets_files.go` may still contain the old assets, right?

The benefit would surely then come from not including the assets in the build, right?

In regards to "bundling goose": I thought about including only the goose "library" and offering commands like `applikatoni migrate` or something, instead of using the goose binary for that. That may be a way to go... 

Also, bundling the templates would be great.

The best case for me would be the following: in development-mode everything works as is but when making a new release "build" everything is packaged into a binary.

**Comment by Wind0r** [comment]
1.) Correct.. If someone change stuff in the assets folder and do not exectue go generate it can happen that old assets are served (if the files on disk are not found). 
Thats why I add go generate to the build/release script. 
For local development I think we can be sure that the assets are on disk.

With "bundling goose" i meant what you descripted =)

I already thought templates are served from binary.
They are bundled with the binary but not used.
I changed it now. If templets not found on disk applikation will use the bundled ones.

**Comment by mrnugget** [comment]
The `build.sh` can be changed to not copy the assets, right?

Do you think this is ready to merge or do you want to give bundling of SQL files a go?

**Comment by Wind0r** [comment]
>  The build.sh can be changed to not copy the assets, right?
So that build.sh do not copy assets to the .tar? Yes, since assets and templates can be loaded from binary.

I think it is ready to merge, but we also can wait until sql files are bundled with the binary.

**Comment by mrnugget** [comment]
Could you remove the copying of the assets from `build.sh` and add yourself to the changelog?

**Comment by mrnugget** [comment]
Besides needing a rebase onto master: do you think this is ready to merge?

**Comment by Wind0r** [comment]
i think so

**Comment by mrnugget** [comment]
Could you rebase this branch onto master so I can merge it? :)

**Comment by Wind0r** [comment]
I will do it after #13 and #16 are merged.. OK?

**Comment by mrnugget** [comment]
Sounds good!

**Comment by Wind0r** [comment]
Assets updated and rebased with current master.

**Comment by mrnugget** [comment]
I checked it out locally and this work beautifully! Thanks a lot for doing this and spending so much time on it.

But I've still got a question: whenever I now want to create a build with `./build.sh` then the `asset_files.go` changes (updated timestamp of the assets in this case). What should we do about that? Make a commit whenever a new version is released?

**Comment by Wind0r** [comment]
No Problem.
mhh. normaly i only would update asset_files when the assets change.
build.sh do not edit assets so there should be any problem or?
if i do a build.sh and a build.sh there isnt any change..
Maybe it happened because i generated the last assets_files with cli>go generate and not with the build script.

**Comment by Wind0r** [comment]
I changed the assets.
I hope ./build.sh will do not change asset_files.go

**Comment by mrnugget** [comment]
It seems like it always updates the `assets_files.go` with new timestamps:

This is the diff after running `go generate`/`build.sh`:
```diff
diff --git a/server/assets_files.go b/server/assets_files.go
index d993c63..a588a85 100644
--- a/server/assets_files.go
+++ b/server/assets_files.go
@@ -99,7 +99,7 @@ func assetsApplikatoniCss() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/applikatoni.css", size: 2101, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/applikatoni.css", size: 2101, mode: os.FileMode(420), modTime: time.Unix(1449122585, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -119,7 +119,7 @@ func assetsApplikatoniJs() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/applikatoni.js", size: 11065, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/applikatoni.js", size: 11065, mode: os.FileMode(420), modTime: time.Unix(1448949305, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -139,7 +139,7 @@ func assetsClicksparkMinJs() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/clickspark.min.js", size: 5163, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/clickspark.min.js", size: 5163, mode: os.FileMode(420), modTime: time.Unix(1440047386, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -159,7 +159,7 @@ func assetsFaviconIco() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/favicon.ico", size: 5430, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/favicon.ico", size: 5430, mode: os.FileMode(420), modTime: time.Unix(1440047386, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -179,7 +179,7 @@ func assetsFaviconJs() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/favicon.js", size: 1394, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/favicon.js", size: 1394, mode: os.FileMode(420), modTime: time.Unix(1440047386, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -199,7 +199,7 @@ func assetsFaviconPng() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/favicon.png", size: 2580, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/favicon.png", size: 2580, mode: os.FileMode(420), modTime: time.Unix(1440047386, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -219,7 +219,7 @@ func assetsLogo_bigPng() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/logo_big.png", size: 32671, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/logo_big.png", size: 32671, mode: os.FileMode(420), modTime: time.Unix(1443876494, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -239,7 +239,7 @@ func assetsLogo_squarePng() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/logo_square.png", size: 66014, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/logo_square.png", size: 66014, mode: os.FileMode(420), modTime: time.Unix(1440047386, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -259,7 +259,7 @@ func assetsTemplatesApplicationTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/application.tmpl", size: 4425, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/application.tmpl", size: 4425, mode: os.FileMode(420), modTime: time.Unix(1443862138, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -279,7 +279,7 @@ func assetsTemplatesDaily_digestTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/daily_digest.tmpl", size: 18732, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/daily_digest.tmpl", size: 18732, mode: os.FileMode(420), modTime: time.Unix(1440229105, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -299,7 +299,7 @@ func assetsTemplatesDeploymentTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/deployment.tmpl", size: 1872, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/deployment.tmpl", size: 1872, mode: os.FileMode(420), modTime: time.Unix(1444145732, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -319,7 +319,7 @@ func assetsTemplatesDeploymentsTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/deployments.tmpl", size: 829, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/deployments.tmpl", size: 829, mode: os.FileMode(420), modTime: time.Unix(1449330023, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -339,7 +339,7 @@ func assetsTemplatesHogan_templatesTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/hogan_templates.tmpl", size: 6393, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/hogan_templates.tmpl", size: 6393, mode: os.FileMode(420), modTime: time.Unix(1448949937, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -359,7 +359,7 @@ func assetsTemplatesHomeTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/home.tmpl", size: 159, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/home.tmpl", size: 159, mode: os.FileMode(420), modTime: time.Unix(1440047386, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -379,7 +379,7 @@ func assetsTemplatesLayoutTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/layout.tmpl", size: 2759, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/layout.tmpl", size: 2759, mode: os.FileMode(420), modTime: time.Unix(1449122585, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -399,7 +399,7 @@ func assetsTemplatesPartialsTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/partials.tmpl", size: 1476, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/partials.tmpl", size: 1476, mode: os.FileMode(420), modTime: time.Unix(1442381956, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
@@ -419,7 +419,7 @@ func assetsTemplatesToni_configurationTmpl() (*asset, error) {
                return nil, err
        }

-       info := bindataFileInfo{name: "assets/templates/toni_configuration.tmpl", size: 301, mode: os.FileMode(420), modTime: time.Unix(1449775235, 0)}
+       info := bindataFileInfo{name: "assets/templates/toni_configuration.tmpl", size: 301, mode: os.FileMode(420), modTime: time.Unix(1440047386, 0)}
        a := &asset{bytes: bytes, info: info}
        return a, nil
 }
```

**Comment by mrnugget** [comment]
It seems like this is related to this issue: https://github.com/jteeuwen/go-bindata/issues/120

**Comment by mrnugget** [comment]
I think the problem is that go-bindata uses the "last modified" timestamp provided by the underlying OS/file system. That changes with every clone of this repository and says nothing about the "last modified" information stored in git.

My local clone of this repository:

```bash
% ls -lrT assets/applikatoni.js
-rw-r--r--  1 mrnugget  staff  11065 Dec  1 06:55:05 2015 assets/applikatoni.js
```

A fresh clone of the repository:

```bash
~/tmp % git clone git@github.com:applikatoni/applikatoni.git
[...]
~/tmp % cd applikatoni/server
~/tmp/applikatoni/server % ls -lrT assets/applikatoni.js
-rw-r--r--  1 mrnugget  staff  11065 Dec 12 14:05:50 2015 assets/applikatoni.js
```

This also explains why the timestamps **decrease** when I check out this branch and run `go generate`: my local clone is older than yours.

I'm not sure whether I like this. I'm still thinking about the pros/cons of this pull request (which has nothing to do with the great work you did!).

I don't see the asset files in the build as a huge minus. At least for the moment.

Including them in the binary would be really neat, but I don't think it would make deployment and so on that much easier. But it adds a new third-party dependency, some "magic" and a potential useless "updated assets_files.go" commit to every release/pull request. 

What do you think about this? Is this something you really want to see merged because you really need it (makes deployment simpler and so on) or just because it's a neat and cool idea (which I can totally understand - it is pretty cool)?

**Comment by Wind0r** [comment]
I'm very sorry for the late reply, I was busy.

Since we still need to ship a tar/zip and we got those problem i have no problems with not getting this pull request accepted.
I am wondering about those file changes and how other project do this. 
I do not realy need it. I saw it in other projects and liked it bause of the single binary deploy.

Maybe the last modified "bug" get fixed soon, then we can watch over it again.

**Comment by mrnugget** [comment]
That sounds good. Let's keep this PR in mind and maybe revisit it some day. Thanks a lot for your work, even though it's not going to get merged now. I know that can be frustrating, but I'm still thankful for you pulling through and finishing this so that we could evaluate the solution properly.

**Pull request closed without merging**